lib/Dialect/AIEVec/Transforms/AIEVectorize.cpp

```diff
@@ -1954,6 +1954,254 @@ static void fuseFMAOpsForColumnTopology(func::FuncOp func, VectState *state) {
+      mulOrFMAOp->getOperand(1) != Op->getOperand(1))
+    return false;
+
+  Value lhs = nullptr;
+  Value rhs = nullptr;
+  Value acc = nullptr;
+  bool isMulOp = false;
+
+  // If the acc operand is a mul op, we will try to generate mul_conv operation
+  // If the acc operand is a fma op, we will try to generate fma_conv operation
+  if (auto mulOp = dyn_cast<aievec::MulOp>(mulOrFMAOp)) {
+    isMulOp = true;
+
+    // Determine the lhs and rhs values for the mul_conv
+    lhs = mulOp->getOperand(0);
+    rhs = mulOp->getOperand(1);
+  } else {
+    auto fmaOp = cast<aievec::FMAOp>(mulOrFMAOp);
+
+    // Determine the lhs, rhs and acc values for the fma_conv
+    lhs = fmaOp->getOperand(0);
+    rhs = fmaOp->getOperand(1);
+    acc = fmaOp->getOperand(2);
+  }
+
+  // Check 6. The def of two operands are upd operations
+  auto lUpdOp = dyn_cast<aievec::UPDOp>(lhs.getDefiningOp());
+  auto rUpdOp = dyn_cast<aievec::UPDOp>(rhs.getDefiningOp());
+
+  if (!lUpdOp || !rUpdOp) {
+    return false;
+  }
+
+  // Check 7. All the ops should belong to the same block, otherwise we might
+  // not be able to fuse them safely
+  if (lhs.getParentBlock() != rhs.getParentBlock())
+    return false;
+
+  if (acc && rhs.getParentBlock() != acc.getParentBlock())
+    return false;
+
+  // Check 8. xstart and zstart distance between two operations should be
+  // 2. offsets, offsets_hi, square and step of two operations should be same.
+  if (isMulOp) {
+    aievec::MulOp defOp = cast<aievec::MulOp>(mulOrFMAOp);
+
+    if (curOp.getOffset(0) != defOp.getOffset(0) ||
+        curOp.getOffsetHi(0) != defOp.getOffsetHi(0) ||
+        curOp.getSquare(0) != defOp.getSquare(0) ||
+        curOp.getStep(0) != defOp.getStep(0) ||
+        curOp.getOffset(1) != defOp.getOffset(1) ||
+        curOp.getOffsetHi(1) != defOp.getOffsetHi(1) ||
+        curOp.getSquare(1) != defOp.getSquare(1) ||
+        curOp.getStep(1) != defOp.getStep(1) ||
+        stoi((std::string)curOp.getStart(0)) -
+                stoi((std::string)defOp.getStart(0)) !=
+            2 ||
+        stoi((std::string)curOp.getStart(1)) -
+                stoi((std::string)defOp.getStart(1)) !=
+            2)
+      return false;
+  } else {
+    aievec::FMAOp defOp = cast<aievec::FMAOp>(mulOrFMAOp);
+    if (curOp.getOffset(0) != defOp.getOffset(0) ||
+        curOp.getOffsetHi(0) != defOp.getOffsetHi(0) ||
+        curOp.getSquare(0) != defOp.getSquare(0) ||
+        curOp.getStep(0) != defOp.getStep(0) ||
+        curOp.getOffset(1) != defOp.getOffset(1) ||
+        curOp.getOffsetHi(1) != defOp.getOffsetHi(1) ||
+        curOp.getSquare(1) != defOp.getSquare(1) ||
+        curOp.getStep(1) != defOp.getStep(1) ||
+        stoi((std::string)curOp.getStart(0)) -
+                stoi((std::string)defOp.getStart(0)) !=
+            2 ||
+        stoi((std::string)curOp.getStart(1)) -
+                stoi((std::string)defOp.getStart(1)) !=
+            2)
+      return false;
+  }
+
+  return true;
+}
```

Comment:
A cleaner way to do this, write a new "satisfiesProperty" (templated) method, and end the method with:

```cpp
return isMulOp && satisfiesProperty(defOp, cast<aievec::MulOp>(mulOrFMAOp)) ||
                  satisfiesProperty(defOp, cast<aievec::FMAOp>(mulOrFMAOp));
```

Not sure how I'd call that property, though. `matchOffsetAndXDistanceForFusion`, `mulOpAttributesAllowFusion`?

Reply:
Thanks! This is a good point!